muspy/processors/event.py

```diff
@@ -0,0 +1,394 @@
+        ignore_empty_tracks: bool = False,
+        resolution: int = DEFAULT_RESOLUTION,
+        force_resolution: bool = False,
+        duplicate_note_mode: str = "fifo",
+    ):
+        self.use_single_note_off_event = use_single_note_off_event
+        self.use_end_of_sequence_event = use_end_of_sequence_event
+        self.encode_velocity = encode_velocity
+        self.force_velocity_event = force_velocity_event
+        self.max_time_shift = max_time_shift
+        self.velocity_bins = velocity_bins
+        self.default_velocity = default_velocity
+        self.encode_instrument = encode_instrument
+        self.default_program = default_program
+        self.default_is_drum = default_is_drum
+        self.num_tracks = num_tracks
+        self.ignore_empty_tracks = ignore_empty_tracks
+        self.resolution = resolution
+        self.force_resolution = force_resolution
+        self.duplicate_note_mode = duplicate_note_mode
+
+        if encode_instrument and num_tracks is None:
+            raise ValueError(
+                'Cannot encode instruments when num_tracks is None')
+
+        # Create vocabulary of events
+        vocab_list: list = []
+        track_ids = [0] if num_tracks is None else range(num_tracks)
+        vocab_list.extend(
+            (NOTE_ON, tr, i)
+            for tr in track_ids
+            for i in range(128))
+        vocab_list.extend(
+            (NOTE_OFF, tr, i)
+            for tr in track_ids
+            for i in (
+                [ALL_NOTES] if use_single_note_off_event else range(128)))
+        vocab_list.extend(
+            (TIME_SHIFT, t) for t in range(1, max_time_shift + 1))
+        if encode_velocity or num_tracks is None:
+            # In single-track mode, always include velocity tokens
+            # for backwards compatibility
+            vocab_list.extend(
+                (VELOCITY, tr, v)
+                for tr in track_ids
+                for v in range(velocity_bins))
+        if encode_instrument:
+            vocab_list.extend((PROGRAM, tr, pr)
+                              for tr in track_ids
+                              for pr in range(128))
+            vocab_list.extend((DRUM, tr) for tr in track_ids)
+        if use_end_of_sequence_event:
+            vocab_list.append((EOS,))
+
+        # Map human-readable tuples to integers
+        self.vocab: frozenbidict[Any, int] = frozenbidict(
+            enumerate(vocab_list)).inverse
+
+    def encode(self, music: Music) -> ndarray:
+        if music.resolution != self.resolution:
+            if self.force_resolution:
+                music = copy.deepcopy(music)
+                music.adjust_resolution(target=self.resolution)
+            else:
+                raise ValueError(
+                    'Expected a resolution of {} TPQN, got {}. '.format(
+                        self.resolution, music.resolution)
+                    + 'Set force_resolution=True to convert the resolution')
```

Comment:
I think warnings work better here. Someone might want to have it that way if, for example, they never want to decode it back.